wagtail/utils/deprecation.py

```diff
@@ -7,3 +11,53 @@ class RemovedInWagtail28Warning(DeprecationWarning):
 
 class RemovedInWagtail29Warning(PendingDeprecationWarning):
     pass
+
+
+class MovedDefinitionHandler(object):
+    """
+    A wrapper for module objects to enable definitions to be moved to a new module, with a
+    deprecation path for the old location. Importing the name from the old location will
+    raise a deprecation warning (but will still complete successfully).
+
+    To use, place the following code in the old module:
+
+    import sys
+    from wagtail.utils.deprecation import MovedDefinitionHandler, RemovedInWagtailXWarning
+
+    MOVED_DEFINITIONS = {
+        'SomeClassOrVariableName': 'path.to.new.module',
+    }
+
+    sys.modules[__name__] = MovedDefinitionHandler(sys.modules[__name__], MOVED_DEFINITIONS, RemovedInWagtailXWarning)
+    """
+    def __init__(self, real_module, moved_definitions, warning_class):
+        self.real_module = real_module
+        self.moved_definitions = moved_definitions
+        self.warning_class = warning_class
+
+    def __getattr__(self, name):
+        try:
+            return getattr(self.real_module, name)
+        except AttributeError as e:
+            try:
+                # is the missing name one of our moved definitions?
+                new_module_name = self.moved_definitions[name]
+            except KeyError:
+                # raise the original AttributeError without including the inner try/catch
+                # in the stack trace
+                raise e from None
+
+            warnings.warn(
+                "%s has been moved from %s to %s" % (name, self.real_module.__name__, new_module_name),
+                category=self.warning_class, stacklevel=2
+            )
+
+            # load the requested definition from the module named in moved_definitions
+            new_module = import_module(new_module_name)
```

Comment:
If there's an exception when importing the relocated module, the stack trace confusingly contains the (handled) exception about the original import:

```
$ python -Wall
>>> from wagtail.admin.utils import get_object_usage
__main__:1: RemovedInWagtail29Warning: get_object_usage has been moved from wagtail.admin.utils to wagtail.admin.models
Traceback (most recent call last):
  File "/Users/chosaka/Projects/wagtail/wagtail/utils/deprecation.py", line 40, in __getattr__
    return getattr(self.real_module, name)
AttributeError: module 'wagtail.admin.utils' has no attribute 'get_object_usage'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/chosaka/Projects/wagtail/wagtail/utils/deprecation.py", line 56, in __getattr__
    new_module = import_module(new_module_name)
   ...
  File "/Users/chosaka/.virtualenvs/testwagtail/lib/python3.6/site-packages/django/conf/__init__.py", line 64, in _setup
    % (desc, ENVIRONMENT_VARIABLE))
django.core.exceptions.ImproperlyConfigured: Requested setting INSTALLED_APPS, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
```

If we were to move these remaining lines out of the `except` clause, [like this](https://gist.github.com/chosak/10d9b58de8d6a1165654ff37bc11437d), that'd be properly hidden:

```
$ python -Wall
Python 3.6.8 (default, May  5 2019, 07:40:18) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from wagtail.admin.utils import get_object_usage
__main__:1: RemovedInWagtail29Warning: get_object_usage has been moved from wagtail.admin.utils to wagtail.admin.models
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/chosaka/Projects/wagtail/wagtail/utils/deprecation.py", line 56, in __getattr__
    new_module = import_module(new_module_name)
   ...
  File "/Users/chosaka/.virtualenvs/testwagtail/lib/python3.6/site-packages/django/conf/__init__.py", line 64, in _setup
    % (desc, ENVIRONMENT_VARIABLE))
django.core.exceptions.ImproperlyConfigured: Requested setting INSTALLED_APPS, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
```

Reply:
Makes sense, cheers! Overly expansive try/catch blocks are never a good thing... fixed in 7cfa1fb8c1b5001fa000a3458b856b1d236187c5.